sources/api/apiclient/src/update.rs

```diff
@@ -0,0 +1,393 @@
+    url: S1,
+    method: S2,
+    data: Option<String>,
+    command_name: &'static str,
+    wait: &WaitPolicy,
+) -> Result<(String, String)>
+where
+    P: AsRef<Path>,
+    S1: AsRef<str>,
+    S2: AsRef<str>,
+{
+    // Fetch the initial status of the API so we know when it's changed.
+    let (code, initial_body) = raw_request_unchecked(&socket_path, "/updates/status", "GET", None)
+        .await
+        .context(error::GetStatus)?;
+
+    // The timestamp is the primary field we use to notice a change.
+    // The first call to the update API on a new system will return a 404, which is fine.
+    let before_timestamp = if code == StatusCode::NOT_FOUND {
+        "first call".to_string()
+    } else if code.is_success() {
+        response_field(&["most_recent_command", "timestamp"], &initial_body)
+            .unwrap_or_else(|| "first call".to_string())
+    } else {
+        return error::MissingStatus {
+            code,
+            body: initial_body,
+        }
+        .fail();
+    };
+    debug!("Found initial timestamp '{}'", before_timestamp);
+
+    // Make the real request the user wanted.
+    let (_code, response_body) = raw_request(&socket_path, &url, &method, data)
+        .await
+        .context(error::Request { command_name })?;
+
+    // Note: we've now made the real request the user asked for, and the rest is our bookkeeping to
+    // wait for it to finish.  We're more careful with retries and don't want to early-exit with ?.
+
+    let mut attempt: u32 = 0;
+    let mut failures: u32 = 0;
+    // How many times we'll retry our bookkeeping checks, e.g. if the status API fails.
+    let max_failures: u32 = 5;
+    // How often to let the user know we're still waiting.
+    let notify_every = Duration::from_secs(5);
+    // Counter so we can tell whether we should notify.
+    let mut waited = Duration::from_millis(0);
+
+    loop {
+        // Check if we've timed out or failed too many requests.
+        attempt += 1;
+        ensure!(
+            attempt < wait.max_attempts,
+            error::TimedOut {
+                waited: format!("{:?}", wait.between_attempts * wait.max_attempts),
+                method: method.as_ref(),
+                url: url.as_ref(),
+            }
+        );
+        ensure!(
+            failures < max_failures,
+            error::StatusCheck {
+                failures,
+                method: method.as_ref(),
+                url: url.as_ref(),
+            }
+        );
+
+        // Let the user know what's going on every once in a while, as we wait.
+        if attempt > 1 && waited >= notify_every {
+            waited = Duration::from_millis(0);
+            info!(
+                "Still waiting for updated status, will wait up to {:?} longer...",
+                (wait.max_attempts * wait.between_attempts) - (attempt * wait.between_attempts)
+            );
+        }
+        thread::sleep(wait.between_attempts);
+        waited += wait.between_attempts;
```

Comment:
I'd prefer to use one of the tokio functions for this, or at least add a warning in the comment that we're doing something that could cause problems down the road if the call pattern changes.